### PR TITLE
CentCom Viewer Admin Panel

### PR DESCRIPTION
**Pull request by Sovexe**
## About the PR 
* Adds a new admin panel that pulls in ban data from CentCom Viewer via its public API
* Includes a Filter Inactive button to hide expired and lifted bans
* Supports both server bans and job bans
* Accessible via the CentCom link in the player options panel

image

image
Filter Inactive mode:
image

## Why's this needed? 
* Allow admins to run a deeper check on a player's background at the click of a button to better handle rule breakers

### CODE REVIEW

File: code/modules/admin/admin.dm [hunk 1]
```diff
@@ -512,6 +512,13 @@ var/global/noir = 0
 
 			return
 
+		if("centombans")
```
Comment (ZeWaka) [inline code/modules/admin/admin.dm:515]: `centom`?

File: code/modules/admin/centcomviewer.dm [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+/datum/centcomviewer/ui_status(mob/user)
+  return tgui_admin_state.can_use_topic(src, user)
+
+/datum/centcomviewer/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if (!ui)
+		ui = new(user, src, "CentComViewer")
+		ui.open()
+	else
+		update_static_data(user, ui)
```
Comment (ZeWaka) [inline code/modules/admin/centcomviewer.dm:21]: what's this for? this shouldn't be needed

File: code/modules/admin/centcomviewer.dm [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+/datum/centcomviewer/ui_static_data(mob/user)
+	var/datum/http_request/request = new()
+	request.prepare(RUSTG_HTTP_METHOD_GET, "https://centcom.melonmesa.com/ban/search/[target_key]", "", "")
+	request.begin_async()
+	UNTIL(request.is_complete())
+	var/datum/http_response/response = request.into_response()
+	. = list(
+		"banData" = response.body,
+		"key" = target_key,
```
Comment (ZeWaka) [inline code/modules/admin/centcomviewer.dm:31]:
```suggestion
			"banData" = response.body,
			"key" = target_key,
```

File: code/modules/admin/centcomviewer.dm [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+		)
+
+/datum/centcomviewer/ui_data(mob/user)
+	. = list(
+		"filterInactive" = filterInactive
+		)
```
Comment (ZeWaka) [inline code/modules/admin/centcomviewer.dm:37]:
```suggestion
	. = list(
			"filterInactive" = filterInactive
		)
```

File: tgui/packages/tgui/interfaces/CentComViewer.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+  // Define a functional component for rendering each ban object.
+  const RenderBans = () => {
+    return banDataJson.map((ban, index) => {
+      // Destructure the properties of the ban object.
+      const { active, bannedOn, expires, jobs, reason, sourceName, type, unbannedBy } = ban;
+      const expired = new Date(expires) < new Date();
+
+      const getBanLength = (bannedOn, expires) => {
+        const ONE_SECOND = 1000;
```
Comment (ZeWaka) [inline tgui/packages/tgui/interfaces/CentComViewer.js:28]: probably move this out of the CentComViewer functional component into just the file scope?

File: tgui/packages/tgui/interfaces/CentComViewer.js [hunk 6]
```diff
@@ -0,0 +1,129 @@
+/**
+ * @file
+ * @copyright 2022
```
Comment (ZeWaka) [inline tgui/packages/tgui/interfaces/CentComViewer.js:3]:
```suggestion
 * @copyright 2023
```

File: code/modules/admin/centcomviewer.dm (hunk 2 above)
Comment (Sovexe) [reply code/modules/admin/centcomviewer.dm:21]: if you already have the panel and then try to open another for a different player it wont fresh unless you hit f5. This forces it to grab the new ban data and target key

**Comment by Sovexe** [comment]
!merge_upstream

**Review by ZeWaka — APPROVED** [review]

**Pull request merged**